core/services/workflows/syncer/handler.go

```diff
@@ -334,103 +334,102 @@ func (w workflowName) Hex() string {
-		BinaryURL:     payload.BinaryURL,
-		ConfigURL:     payload.ConfigURL,
-	}
+		status := job.WorkflowSpecStatusActive
+		if payload.Status == 1 {
+			status = job.WorkflowSpecStatusPaused
+		}
 
-	if _, err = h.workflowArtifactsStore.UpsertWorkflowSpecWithSecrets(ctx, entry, payload.SecretsURL, hex.EncodeToString(urlHash), string(secrets)); err != nil {
-		return fmt.Errorf("failed to upsert workflow spec with secrets: %w", err)
-	}
+		// Create a new entry in the workflow_spec table corresponding for the new workflow, with the contents of the binaryURL + configURL in the table
+		entry := &job.WorkflowSpec{
+			Workflow:      hex.EncodeToString(decodedBinary),
+			Config:        string(config),
+			WorkflowID:    wfID,
+			Status:        status,
+			WorkflowOwner: owner,
+			WorkflowName:  payload.WorkflowName,
+			SpecType:      job.WASMFile,
+			BinaryURL:     payload.BinaryURL,
+			ConfigURL:     payload.ConfigURL,
+		}
 
-	if status != job.WorkflowSpecStatusActive {
-		h.lggr.Debugw("workflow is marked as paused, so not starting it", "workflow", wfID)
-		return nil
+		if _, err = h.workflowArtifactsStore.UpsertWorkflowSpecWithSecrets(ctx, entry, payload.SecretsURL, hex.EncodeToString(urlHash), string(secrets)); err != nil {
+			return fmt.Errorf("failed to upsert workflow spec with secrets: %w", err)
+		}
 	}
 
-	// If status == active, start a new WorkflowEngine instance, and add it to local engine registry
-	engine, err := h.engineFactory(
-		ctx,
-		wfID,
-		owner,
-		workflowName{
-			name: payload.WorkflowName,
-		},
-		config,
-		decodedBinary,
-	)
-	if err != nil {
-		return fmt.Errorf("failed to create workflow engine: %w", err)
-	}
+	// If there is no engine in the engine registry, then create it, start it, and register it
+	engineKey := EngineRegistryKey{Owner: payload.WorkflowOwner, Name: payload.WorkflowName}
+	if !h.engineRegistry.Contains(engineKey) {
+		decodedBinary, err := hex.DecodeString(spec.Workflow)
+		if err != nil {
+			return fmt.Errorf("failed to create workflow engine: %w", err)
+		}
+		// Start a new WorkflowEngine instance, and add it to local engine registry
+		engine, err := h.engineFactory(
+			ctx,
+			wfID,
+			owner,
+			workflowName{
+				name: payload.WorkflowName,
+			},
+			[]byte(spec.Config),
+			decodedBinary,
+		)
+		if err != nil {
+			return fmt.Errorf("failed to create workflow engine: %w", err)
+		}
 
-	if err := engine.Start(ctx); err != nil {
-		return fmt.Errorf("failed to start workflow engine: %w", err)
-	}
+		if err := engine.Start(ctx); err != nil {
+			return fmt.Errorf("failed to start workflow engine: %w", err)
+		}
 
-	if err := h.engineRegistry.Add(key, engine, payload.WorkflowID); err != nil {
-		// This shouldn't happen because we call the handler serially and
-		// check for running engines above, see the call to engineRegistry.Contains.
-		return fmt.Errorf("invariant violation: %w", err)
+		if err := h.engineRegistry.Add(engineKey, engine, payload.WorkflowID); err != nil {
```

Comment:
@justinkaseman We should also clean up after ourselves here by calling engine.Close